internal/serialization/portable_context.go

```diff
@@ -110,19 +114,30 @@ func (c *PortableContext) LookUpOrRegisterClassDefiniton(portable serialization.
 }
 
 func (c *PortableContext) LookUpClassDefinition(factoryID int32, classID int32, version int32) *serialization.ClassDefinition {
+	c.mu.RLock()
 	factory := c.classDefContext[factoryID]
+	c.mu.RUnlock()
 	if factory == nil {
 		return nil
 	}
 	return factory.LookUp(classID, version)
 }
 
-func (c *PortableContext) RegisterClassDefinition(classDefinition *serialization.ClassDefinition) error {
-	factoryID := classDefinition.FactoryID
-	if c.classDefContext[factoryID] == nil {
-		c.classDefContext[factoryID] = NewClassDefinitionContext(factoryID)
+func (c *PortableContext) RegisterClassDefinition(cd *serialization.ClassDefinition) error {
+	c.mu.RLock()
+	factoryID := cd.FactoryID
+	f := c.classDefContext[factoryID]
+	c.mu.RUnlock()
```

Comment:
What about `defer c.mu.RUnlock()` after line 127 instead of locking again if `f==nil`?

Reply:
We cannot lock the mutex for writing while it it is locked for reading. We could use a `sync.Mutex` instead of `sync.RWMutex` at the start of the method and defer unlocking it. But in our use case, we would mostly reading the map and seldomly updating it. Since the mutex can be read locked concurrently but can be write locked only one goroutine, using `sync.Mutex, would create a bottleneck for all  goroutines. So, it is better to use `sync.RWMutex` to check the existence of the value we're looking for.

Reply:
Thanks for explanation.